### DISCUSSION TITLE
How to mock `msg.sender.call`?

### DISCUSSION THREAD

**Discussion by 0xCLARITY** [foundry-rs/foundry#2488] [Help] [answered]
I have a 721 mint function for a free mint that attempts to refund any excess balance sent:

```
    error RefundFailed(address to, uint256 amount);
        
    function mint() external payable {
        _mint(msg.sender, 1);
        
        // Refund user if too much ETH was accidentally sent for minting.
        if (msg.value > 0) {
            (bool success, ) = msg.sender.call{value: msg.value}("");
            if (!success) revert RefundFailed(msg.sender, msg.value);
        }
    }
```

So, I thought, I would do something like this in my test file:
```
    function test_mint_refund_failure() external {
        address alice = address(128);
        vm.deal(alice, 10**18);

        vm.startPrank(alice, alice);
        vm.mockCall(alice, abi.encodeWithSelector(alice.call.selector), abi.encode(false, ""));
        vm.expectRevert(RefundFailed.selector);
        contract.mint{value: 100}();
    }
```

But, for reasons slightly outside of my understanding of Solidity / the EVM - `alice.call` does not have a function selector.

How _would_ I go about mocking the call or manipulating the VM state such that I can hit my `revert RefundFailed()` line?

**Comment by onbjerg** [comment, marked as answer]
`address.call("")` just invokes the `CALL` opcode with no calldata, so you do not need to use `vm.mockCall`. Although, in order to hit your path, you would essentially need to create a contract with a payable `receive` function that reverts. Returning `false` does not do anything - the `success` boolean just indicates whether the `CALL` opcode encountered a revert or not.
